developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,255 @@
+
+When running natively, the SOF uses the ALSA topology API to configure audio
+pipelines on the DSP. The SOF driver reads a platform-specific topology file
+from the filesystem, parses it and sends all the components from it to the DSP
+firmware.
+
+When configured to support virtualised audio, the host topology additionally
+provides connection points for guest audio pipelines. Guest playback pipelines
+connect to mixer and capture pipelines to demultiplexer components. The ALSA /
+ASoC / DAPM subsystems on the host are unaware of any guests, but they should be
+aware of any streaming, taking place on host parts of VM pipelines for
+configuration and power management. That means, e.g. if a guest initialises
+playback, the ASoC subsystem on the host should be aware, that the pipeline,
+connecting the respective mixer and the used codec is currently performing audio
+playback. To achieve this, for each such guest connection point we add a virtual
+PCM as a widget of type "aif_in" or "aif_out" for playback and capture
+respectively, and a virtual 0-size buffer. They are only visible on the host and
+are not sent to the DSP.
+
+When running in a VM, the SOF also uses ALSA topology, but in this case the
+topology should have no relation to the host hardware. The guest audio topology
+should be decided upon by the host system administrator. Therefore we let the
+SOF instance, running on the guest, obtain topology from the host. Such a
+topology should only contain "software components," like PCMs, buffers, software
+volume controls. Those components thus represent partial pipelines, that then
+will be linked to mixer and demux components of the host SOF topology for
+playback and capture respectively.
+
+With this design the topology is represented in the following way:
+
+DSP: has a full topology, including host and guests' components. No distinction
+is made between the host and any guests.
+
+host: the SOF driver core, running on the host, and the ASoC subsystem only see
+the host topology. The vhost driver manages connection points between the host
+and the guests' topology fragments.
+
+guest: only "sees" the guest parts.
+
+As mentioned above, SOF IPC messages from guests are forwarded exactly 1-to-1 to
+the DSP. E.g., if a guest sends an IPC to adjust audio volume, it will be
+forwarded to the DSP. That IPC contains the volume component ID as seen by the
+respective guest. Therefore it should be exactly the same ID, as the one used by
+the firmware.
+
+3. Implementation
+*****************
+
+3.1. Initialisation
+-------------------
+
+The SOF driver probing on the host remains unmodified, except additionally the
+vhost driver is initialised, which registers a /dev/vhost-dsp misc device.
+
+When QEMU starts a guest instance, it detects the above misc device and
+creates a PCI device for the guest with dedicated vendor and device IDs.
+
+The SOF VirtIO driver on the guest registers support for that PCI device and
+its .probe() method is called.
```

Comment:
what PCI ID is this referring to? Is this the same as the cAVS IP (00:1f.3:) or a new one specific for virtualization?

Reply:
it's a dedicated virtualisation ID from RedHat